openapi/cue/genoapi.go

```diff
@@ -0,0 +1,448 @@
+// of CUE constraints. This step validates the combined sources for consistency.
+//
+//
+// 3. Converting CUE to OpenAPI
+//
+// In this step a self-contained OpenAPI definition is generated for each
+// directory containing proto definitions. Files are made self-contained by
+// including a schema definition for each imported type within the OpenAPI
+// spec itself. To avoid name collissions, types are, by convention, prefixed
+// with their proto package name.
+//
+//
+// Possible extensions to the generation pipeline
+//
+// The generation pipeline can be augmented by injecting CUE from other sources
+// before step 2. As combining CUE sources is a commutative operation, order
+// of injection does not matter and there is no need for the user to be
+// explicit about any order or injection points.
+//
+// Examples of other possible CUE sources are:
+// - hand-written .cue files in each of the cue directories
+// - constraints extracted from Go code
+//
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"cuelang.org/go/cue"
+	"cuelang.org/go/cue/errors"
+	"cuelang.org/go/cue/format"
+	"cuelang.org/go/cue/load"
+	"cuelang.org/go/encoding/openapi"
+	"cuelang.org/go/encoding/protobuf"
+	"github.com/emicklei/proto"
+	"github.com/getkin/kin-openapi/openapi3"
+)
+
+var (
+	inplace = flag.Bool("inplace", false, "generate configurations in place")
+	paths   = flag.String("paths", "/protobuf", "comma-separated path to search for .proto imports")
+	root    = flag.String("root", "", "the istio/api root for which to generate files (default is cwd)")
+	verbose = flag.Bool("v", false, "print debugging output")
+
+	// manually configuring builds
+	outdir  = flag.String("dir", "", "only generated files in this directory")
+	perFile = flag.Bool("per_file", false, "generate separate output per file; dir must be specfied")
+	all     = flag.String("all", "", "combine all the matched outputs in a single file")
+	title   = flag.String("title", "All Istio types.", "title of the output file when --all is used")
+	version = flag.String("version", "v1alpha1", "version of the output file when --all is used")
+)
+
+func main() {
+	flag.Parse()
+	log.SetFlags(log.Lshortfile)
+
+	cwd, _ := os.Getwd()
+
+	// Include paths for proto evaluation. Take relative paths relative to the
+	// current working directory.
+	importPaths := []string{cwd} // change cwd later if root is set.
+	for _, p := range strings.Split(*paths, ",") {
+		if p != "" {
+			abs, _ := filepath.Abs(p)
+			importPaths = append(importPaths, abs)
+		}
+	}
+
+	// Update cwd to istio/api path, if not the actual cwd.
+	if *root != "" {
+		var err error
+		cwd, err = filepath.Abs(*root)
+		if err != nil {
+			log.Fatalf("Invalid root %q: %v", *root, err)
+		}
+		importPaths[0] = cwd
+	}
+
+	b := protobuf.NewExtractor(&protobuf.Config{
+		Root:   cwd,
+		Module: "istio.io/api",
```

Comment:
Is this tool really specific to the Istio use case, or could this be generalized a bit so all Istio-isms are moved out of the tool and into config instead?

Reply:
I consider this to be more of a script than a tool. The cue-openapi API is generic, but this code indeed is pretty specific to Istio as it stands. It includes support for very specific different modes (splitting across different files in the same directory, as well as aggregating types of all directories into a single file, for instance) as well as quite specific documentation and title extraction code.

Some code is related to Istio picking proto naming as as basis for OpenAPI naming. This arguably is not fully specific to Istio and could be hoisted into a utility package.

Also some of the logic could be hoisted as higher-level support into the API. I could probably reduce it by about 150-200 lines hoisting code for the last two cases and another big chunk if the OpenAPI file boundaries were to coincide with the package structure.

The Root and module in your highlighted example could be derived from a cue.mod file or config file if we were to define one.

All in all, this script is designed to figure out Istio's requirements for the OpenAPI generation and CUE constraints and was not meant to be a generic tool. I figured it is likely there will be other Istio-specific requests as well.

Reply:
I think we can go with an istio-specific tool for now and we can extract it to be more generic if needed.